File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/dialogs/EditEventDialog.java
package info.nightscout.androidaps.plugins.general.automation.dialogs;

import android.app.Activity;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.textfield.TextInputEditText;
import com.squareup.otto.Subscribe;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import butterknife.Unbinder;
import info.nightscout.androidaps.MainApp;
import info.nightscout.androidaps.R;
import info.nightscout.androidaps.plugins.general.automation.AutomationEvent;
import info.nightscout.androidaps.plugins.general.automation.AutomationPlugin;
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationAddAction;
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationDataChanged;
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationUpdateGui;
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationUpdateTrigger;
import info.nightscout.androidaps.plugins.general.automation.triggers.TriggerConnector;

public class EditEventDialog extends DialogFragment {
    public interface OnClickListener {
        void onClick(AutomationEvent event);
    }

    private static AutomationEvent staticEvent;

    @BindView(R.id.inputEventTitle)
    TextInputEditText mEditEventTitle;

    @BindView(R.id.editTrigger)
    TextView mEditTrigger;

    @BindView(R.id.addAction)
    TextView mAddAction;

    @BindView(R.id.triggerDescription)
    TextView mTriggerDescription;

    @BindView(R.id.forcedtriggerdescription)
    TextView mForcedTriggerDescription;

    @BindView(R.id.forcedtriggerdescriptionlabel)
    TextView mForcedTriggerDescriptionLabel;

    @BindView(R.id.actionListView)
    RecyclerView mActionListView;

    private Unbinder mUnbinder;
    private ActionListAdapter mActionListAdapter;
    private AutomationEvent mEvent;
    private boolean mAddNew;

    public static EditEventDialog newInstance(AutomationEvent event, boolean addNew) {
        staticEvent = event;

        Bundle args = new Bundle();
        EditEventDialog fragment = new EditEventDialog();
        fragment.setArguments(args);
        // clone event
        try {
            fragment.mEvent = new AutomationEvent().fromJSON(event.toJSON());
        } catch (Exception e) {
            e.printStackTrace();
        }
        fragment.mAddNew = addNew;
        return fragment;
    }

    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.automation_dialog_event, container, false);
        mUnbinder = ButterKnife.bind(this, view);

        // load data from bundle
        if (savedInstanceState != null) {
            String eventData = savedInstanceState.getString("event");
            if (eventData != null) mEvent = new AutomationEvent().fromJSON(eventData);
            mAddNew = savedInstanceState.getBoolean("addNew");
        } else if (mAddNew) {
            mEvent.setTrigger(new TriggerConnector(TriggerConnector.Type.OR));

        }

        // event title
        mEditEventTitle.setText(mEvent.getTitle());

        // display root trigger
        mTriggerDescription.setText(mEvent.getTrigger().friendlyDescription());

        mEditTrigger.setOnClickListener(v -> {
            EditTriggerDialog dialog = EditTriggerDialog.newInstance(mEvent.getTrigger());
            if (getFragmentManager() != null)
                dialog.show(getFragmentManager(), "EditTriggerDialog");
        });

        // setup action list view
        mActionListAdapter = new ActionListAdapter(getFragmentManager(), mEvent.getActions());
        mActionListView.setLayoutManager(new LinearLayoutManager(getContext()));
        mActionListView.setAdapter(mActionListAdapter);

        mAddAction.setOnClickListener(v -> {
            FragmentManager fragmentManager = getFragmentManager();
            if (fragmentManager != null)
                new ChooseActionDialog().show(fragmentManager, "ChooseActionDialog");
        });

        showPreconditions();

        MainApp.bus().register(this);

        return view;
    }

    @Override
    public void onDestroyView() {
        MainApp.bus().unregister(this);
        mUnbinder.unbind();
        super.onDestroyView();
    }

    @Subscribe
    public void onEventAutomationUpdateGui(EventAutomationUpdateGui ignored) {
        Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(() -> {
                mActionListAdapter.notifyDataSetChanged();
                showPreconditions();
            });
    }

    @Subscribe
    public void onEventAutomationAddAction(EventAutomationAddAction event) {
        Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(() -> {
                mEvent.addAction(event.getAction());
                mActionListAdapter.notifyDataSetChanged();
            });
    }

    @Subscribe
    public void onEventAutomationUpdateTrigger(EventAutomationUpdateTrigger event) {
        Activity activity = getActivity();
        if (activity != null)
            activity.runOnUiThread(() -> {
                mEvent.setTrigger(event.getTrigger());
                mTriggerDescription.setText(mEvent.getTrigger().friendlyDescription());
            });
    }

    @OnClick(R.id.ok)
    public void onButtonOk(View unused) {
        // check for title
        String title = mEditEventTitle.getText().toString();
        if (title.isEmpty()) {
            Toast.makeText(getContext(), R.string.automation_missing_task_name, Toast.LENGTH_LONG).show();
            return;
        }
        mEvent.setTitle(title);

        // check for at least one trigger
        TriggerConnector con = (TriggerConnector) mEvent.getTrigger();
        if (con.size() == 0) {
            Toast.makeText(getContext(), R.string.automation_missing_trigger, Toast.LENGTH_LONG).show();
            return;
        }

        // check for at least one action
        if (mEvent.getActions().isEmpty()) {
            Toast.makeText(getContext(), R.string.automation_missing_action, Toast.LENGTH_LONG).show();
            return;
        }

        // apply changes
        staticEvent.fromJSON(mEvent.toJSON());

        // add new
        if (mAddNew) {
            final AutomationPlugin plugin = AutomationPlugin.getPlugin();
            plugin.getAutomationEvents().add(mEvent);
        }

        dismiss();
        MainApp.bus().post(new EventAutomationDataChanged());
    }

    @OnClick(R.id.cancel)
    public void onButtonCancel(View unused) {
        dismiss();
    }

    @Override
    public void onSaveInstanceState(Bundle bundle) {
        bundle.putString("event", mEvent.toJSON());
        bundle.putBoolean("addNew", mAddNew);
    }

    private void showPreconditions() {
        TriggerConnector forcedTriggers = mEvent.getPreconditions();
        if (forcedTriggers.size() > 0) {
            mForcedTriggerDescription.setVisibility(View.VISIBLE);
            mForcedTriggerDescriptionLabel.setVisibility(View.VISIBLE);
            mForcedTriggerDescription.setText(forcedTriggers.friendlyDescription());
        } else {
            mForcedTriggerDescription.setVisibility(View.GONE);
            mForcedTriggerDescriptionLabel.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/info/nightscout/androidaps/plugins/general/automation/dialogs/EditEventDialog.kt
package info.nightscout.androidaps.plugins.general.automation.dialogs

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import info.nightscout.androidaps.MainApp
import info.nightscout.androidaps.R
import info.nightscout.androidaps.databinding.AutomationDialogEventBinding
import info.nightscout.androidaps.dialogs.DialogFragmentWithDate
import info.nightscout.androidaps.plugins.bus.RxBusWrapper
import info.nightscout.androidaps.plugins.general.automation.AutomationEvent
import info.nightscout.androidaps.plugins.general.automation.AutomationPlugin
import info.nightscout.androidaps.plugins.general.automation.actions.Action
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationAddAction
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationDataChanged
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationUpdateAction
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationUpdateGui
import info.nightscout.androidaps.plugins.general.automation.events.EventAutomationUpdateTrigger
import info.nightscout.androidaps.plugins.general.automation.triggers.TriggerConnector
import info.nightscout.androidaps.utils.FabricPrivacy
import info.nightscout.androidaps.utils.ToastUtils
import info.nightscout.androidaps.utils.extensions.plusAssign
import info.nightscout.androidaps.utils.extensions.toVisibility
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import javax.inject.Inject

class EditEventDialog : DialogFragmentWithDate() {

    @Inject lateinit var rxBus: RxBusWrapper
    @Inject lateinit var mainApp: MainApp
    @Inject lateinit var fabricPrivacy: FabricPrivacy
    @Inject lateinit var automationPlugin: AutomationPlugin

    private var actionListAdapter: ActionListAdapter? = null
    private lateinit var event: AutomationEvent
    private var position: Int = -1

    private var disposable: CompositeDisposable = CompositeDisposable()

    private var _binding: AutomationDialogEventBinding? = null

    // This property is only valid between onCreateView and
    // onDestroyView.
    private val binding get() = _binding!!

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View {
        event = AutomationEvent(mainApp)
        // load data from bundle
        (savedInstanceState ?: arguments)?.let { bundle ->
            position = bundle.getInt("position", -1)
            bundle.getString("event")?.let { event = AutomationEvent(mainApp).fromJSON(it) }
        }

        onCreateViewGeneral()
        _binding = AutomationDialogEventBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.okcancel.ok.visibility = (!event.readOnly).toVisibility()

        binding.inputEventTitle.setText(event.title)
        binding.inputEventTitle.isFocusable = !event.readOnly
        binding.triggerDescription.text = event.trigger.friendlyDescription()

        binding.editTrigger.visibility = (!event.readOnly).toVisibility()
        binding.editTrigger.setOnClickListener {
            val args = Bundle()
            args.putString("trigger", event.trigger.toJSON())
            val dialog = EditTriggerDialog()
            dialog.arguments = args
            dialog.show(childFragmentManager, "EditTriggerDialog")
        }

        // setup action list view
        actionListAdapter = ActionListAdapter()
        binding.actionListView.layoutManager = LinearLayoutManager(context)
        binding.actionListView.adapter = actionListAdapter

        binding.addAction.visibility = (!event.readOnly).toVisibility()
        binding.addAction.setOnClickListener { ChooseActionDialog().show(childFragmentManager, "ChooseActionDialog") }

        showPreconditions()

        disposable += rxBus
            .toObservable(EventAutomationUpdateGui::class.java)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                actionListAdapter?.notifyDataSetChanged()
                showPreconditions()
            }, fabricPrivacy::logException)
        disposable += rxBus
            .toObservable(EventAutomationAddAction::class.java)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                event.addAction(it.action)
                actionListAdapter?.notifyDataSetChanged()
            }, fabricPrivacy::logException)
        disposable += rxBus
            .toObservable(EventAutomationUpdateTrigger::class.java)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                event.trigger = it.trigger
                binding.triggerDescription.text = event.trigger.friendlyDescription()
            }, fabricPrivacy::logException)
        disposable += rxBus
            .toObservable(EventAutomationUpdateAction::class.java)
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe({
                event.actions[it.position] = it.action
                actionListAdapter?.notifyDataSetChanged()
            }, fabricPrivacy::logException)
    }

    override fun submit(): Boolean {
        // check for title
        val title = binding.inputEventTitle.text?.toString() ?: return false
        if (title.isEmpty()) {
            ToastUtils.showToastInUiThread(context, R.string.automation_missing_task_name)
            return false
        }
        event.title = title
        // check for at least one trigger
        val con = event.trigger as TriggerConnector
        if (con.size() == 0) {
            ToastUtils.showToastInUiThread(context, R.string.automation_missing_trigger)
            return false
        }
        // check for at least one action
        if (event.actions.isEmpty()) {
            ToastUtils.showToastInUiThread(context, R.string.automation_missing_action)
            return false
        }
        // store
        if (position == -1)
            automationPlugin.add(event)
        else
            automationPlugin.set(event, position)

        rxBus.send(EventAutomationDataChanged())
        return true
    }

    override fun onDestroyView() {
        super.onDestroyView()
        disposable.clear()
        _binding = null
    }

    override fun onSaveInstanceState(savedInstanceState: Bundle) {
        super.onSaveInstanceState(savedInstanceState)
        savedInstanceState.putString("event", event.toJSON())
        savedInstanceState.putInt("position", position)
    }

    private fun showPreconditions() {
        val forcedTriggers = event.getPreconditions()
        if (forcedTriggers.size() > 0) {
            binding.forcedTriggerDescription.visibility = View.VISIBLE
            binding.forcedTriggerDescriptionLabel.visibility = View.VISIBLE
            binding.forcedTriggerDescription.text = forcedTriggers.friendlyDescription()
        } else {
            binding.forcedTriggerDescription.visibility = View.GONE
            binding.forcedTriggerDescriptionLabel.visibility = View.GONE
        }
    }

    inner class ActionListAdapter : RecyclerView.Adapter<ActionListAdapter.ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            val v = LayoutInflater.from(parent.context).inflate(R.layout.automation_action_item, parent, false)
            return ViewHolder(v)
        }

        override fun onBindViewHolder(holder: ViewHolder, position: Int) {
            val action = event.actions[position]
            holder.bind(action, this, position)
        }

        override fun getItemCount(): Int = event.actions.size

        inner class ViewHolder(val view: View) : RecyclerView.ViewHolder(view) {

            fun bind(action: Action, recyclerView: RecyclerView.Adapter<ViewHolder>, position: Int) {
                if (!event.readOnly)
                    view.findViewById<LinearLayout>(R.id.automation_layoutText).setOnClickListener {
                        if (action.hasDialog()) {
                            val args = Bundle()
                            args.putInt("actionPosition", position)
                            args.putString("action", action.toJSON())
                            val dialog = EditActionDialog()
                            dialog.arguments = args
                            dialog.show(childFragmentManager, "EditActionDialog")
                        }
                    }
                view.findViewById<ImageView>(R.id.automation_iconTrash).run {
                    visibility = (!event.readOnly).toVisibility()
                    setOnClickListener {
                        event.actions.remove(action)
                        recyclerView.notifyDataSetChanged()
                        rxBus.send(EventAutomationUpdateGui())
                    }
                }
                view.findViewById<ImageView>(R.id.automation_action_image).setImageResource(action.icon())
                view.findViewById<TextView>(R.id.automation_viewActionTitle).text = action.shortDescription()
            }
        }
    }
}
